libraries/chain/include/eosio/chain/block.hpp

```diff
@@ -85,21 +85,28 @@ namespace eosio { namespace chain {
 
    using block_extension = block_extension_types::block_extension_t;
 
+   using signed_block_ptr = std::shared_ptr<const signed_block>;
+   // mutable_block_ptr is built up until it is signed and converted to signed_block_ptr
+   // mutable_block_ptr is not thread safe and should be moved into signed_block_ptr when complete
+   using mutable_block_ptr = std::unique_ptr<signed_block>;
+
    /**
     */
    struct signed_block : public signed_block_header{
    private:
       signed_block( const signed_block& ) = default;
+      explicit signed_block( const signed_block_header& h ):signed_block_header(h){}
    public:
       signed_block() = default;
-      explicit signed_block( const signed_block_header& h ):signed_block_header(h){}
       signed_block( signed_block&& ) = default;
       signed_block& operator=(const signed_block&) = delete;
       signed_block& operator=(signed_block&&) = default;
-      signed_block clone() const { return *this; }
+      mutable_block_ptr clone() const { return std::unique_ptr<signed_block>(new signed_block(*this)); }
+      static mutable_block_ptr create_mutable_block(const signed_block_header& h) { return std::unique_ptr<signed_block>(new signed_block(h)); }
+      static signed_block_ptr  create_signed_block(mutable_block_ptr&& b) { b->pack(); return signed_block_ptr{std::move(b)}; }
```

Comment:
I don't see why these should be inside `struct signed_block`. even if `pack()`  has tobe made public.

It is also a little strange to me that `bytes packed_block;` is inside `signed_block`? Why not have a separate type `packed_signed_block`? Maybe it would require more changes?

Reply:
They are constructor like methods. Seems like they belong in `signed_block`, also allows `pack()` to be private without adding friends of the external functions.

We emit/signal `signed_block` out to plugins. `signed_block` is a well known and used type. I see this as an internal cached optimization to make `packed_signed_block()` efficient. Creating a different type means that we have to pass around a `packed_signed_block` everywhere we use `signed_block` today.